[WebGPU] Clamp depth values to the viewport range

mwyrzykowski opened the pull request:
#### d8d655468dab980b4a4d83a5e02dff57845f1e95
<pre>
[WebGPU] Clamp depth values to the viewport range
<a href="https://bugs.webkit.org/show_bug.cgi?id=264449">https://bugs.webkit.org/show_bug.cgi?id=264449</a>
&lt;radar://118144869&gt;

Reviewed by Tadeu Zagallo.

Add passing expectations for api,operation,rendering,depth_clip_clamp
and add depth clamping when the viewport is not the default [0, 1] range.

* LayoutTests/http/tests/webgpu/webgpu/api/operation/rendering/depth_clip_clamp-expected.txt:
* Source/WebGPU/WGSL/GlobalVariableRewriter.cpp:
(WGSL::RewriteGlobalVariables::insertParameters):
* Source/WebGPU/WGSL/Metal/MetalFunctionWriter.cpp:
(WGSL::Metal::FunctionDefinitionWriter::visit):
* Source/WebGPU/WebGPU/CommandEncoder.h:
* Source/WebGPU/WebGPU/CommandEncoder.mm:
(WebGPU::CommandEncoder::runDepthClampEncoder):
(WebGPU::CommandEncoder::beginRenderPass):
* Source/WebGPU/WebGPU/RenderPassEncoder.h:
* Source/WebGPU/WebGPU/RenderPassEncoder.mm:
(WebGPU::RenderPassEncoder::RenderPassEncoder):
(WebGPU::RenderPassEncoder::quantizedDepthValue):
(WebGPU::RenderPassEncoder::setPipeline):
(WebGPU::RenderPassEncoder::setViewport):

Canonical link: <a href="https://commits.webkit.org/274666@main">https://commits.webkit.org/274666@main</a>
</pre>
<!--EWS-Status-Bubble-Start-->
https://github.com/WebKit/WebKit/commit/f3d63d10b53d7e1a42c17d1d05ea9ad9ab60c2aa

| Misc | iOS, tvOS & watchOS  | macOS  | Linux |  Windows |
| ----- | ---------------------- | ------- |  ----- |  --------- |
| [✅ 🧪 style](https://ews-build.webkit.org/#/builders/38/builds/39468 "Passed style check") | [✅ 🛠 ios](https://ews-build.webkit.org/#/builders/48/builds/18447 "Built successfully") | [✅ 🛠 mac](https://ews-build.webkit.org/#/builders/14/builds/41822 "Built successfully") | [✅ 🛠 wpe](https://ews-build.webkit.org/#/builders/5/builds/42002 "Built successfully") | [✅ 🛠 wincairo](https://ews-build.webkit.org/#/builders/32/builds/35368 "Built successfully") 
| | [✅ 🛠 ios-sim](https://ews-build.webkit.org/#/builders/49/builds/21326 "Built successfully") | [✅ 🛠 mac-AS-debug](https://ews-build.webkit.org/#/builders/51/builds/15776 "Built successfully") | [✅ 🧪 wpe-wk2](https://ews-build.webkit.org/#/builders/34/builds/32997 "Passed tests") | 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/11/builds/40042 "Passed tests") | [✅ 🧪 ios-wk2](https://ews-build.webkit.org/#/builders/47/builds/15541 "Passed tests") | [✅ 🧪 api-mac](https://ews-build.webkit.org/#/builders/18/builds/34182 "Passed tests") | [✅ 🧪 api-wpe](https://ews-build.webkit.org/#/builders/41/builds/13509 "Passed tests") | 
| | [✅ 🧪 ios-wk2-wpt](https://ews-build.webkit.org/#/builders/42/builds/13472 "Passed tests") | [✅ 🧪 mac-wk1](https://ews-build.webkit.org/#/builders/10/builds/35130 "Passed tests") | [✅ 🛠 gtk](https://ews-build.webkit.org/#/builders/2/builds/43280 "Built successfully") | 
| | [✅ 🧪 api-ios](https://ews-build.webkit.org/#/builders/13/builds/35834 "Passed tests") | [✅ 🧪 mac-wk2](https://ews-build.webkit.org/#/builders/36/builds/35461 "Passed tests") | [✅ 🧪 gtk-wk2](https://ews-build.webkit.org/#/builders/1/builds/39267 "Passed tests") | 
| | [✅ 🛠 tv](https://ews-build.webkit.org/#/builders/44/builds/14280 "Built successfully") | [✅ 🧪 mac-AS-debug-wk2](https://ews-build.webkit.org/#/builders/50/builds/11774 "Passed tests") | [✅ 🧪 api-gtk](https://ews-build.webkit.org/#/builders/21/builds/37529 "Passed tests") | 
| | [✅ 🛠 tv-sim](https://ews-build.webkit.org/#/builders/46/builds/15886 "Built successfully") | | | 
| [✅ 🛠 🧪 merge](https://ews-build.webkit.org/#/builders/19/builds/8895 "Built successfully and passed tests") | [✅ 🛠 watch](https://ews-build.webkit.org/#/builders/43/builds/15934 "Built successfully") | | | 
| | [✅ 🛠 watch-sim](https://ews-build.webkit.org/#/builders/45/builds/15543 "Built successfully") | | | 
<!--EWS-Status-Bubble-End-->

mwyrzykowski commented on Source/WebGPU/WGSL/GlobalVariableRewriter.cpp:
```diff
@@ -1554,7 +1554,7 @@ void RewriteGlobalVariables::insertParameters(AST::Function& function, const Vec
     for (auto group : groups)
         insertParameter(group, argumentBufferParameterName(group));
-    if (!m_globalsUsingDynamicOffset.isEmpty()) {
+    if (!m_globalsUsingDynamicOffset.isEmpty() || (m_stage == ShaderStage::Fragment && m_callGraph.ast().usesFragDepth())) {
```
this emits `__DynamicOffsets` buffer even when dynamic offsets are not used. It is because we pack the min / max depth into `__DynamicOffsets`

mwyrzykowski commented on Source/WebGPU/WGSL/Metal/MetalFunctionWriter.cpp:
```diff
@@ -133,6 +133,10 @@ class FunctionDefinitionWriter : public AST::Visitor {
     Indentation<4> m_indent { 0 };
     std::optional<AST::StructureRole> m_structRole;
     std::optional<ShaderStage> m_entryPointStage;
+    std::optional<ShaderStage> m_cachedEntryPointStage;
```
Maybe there is a much better way to achieve this, I defer to @tadeuzagallo

mwyrzykowski commented on Source/WebGPU/WGSL/Metal/MetalFunctionWriter.cpp:
```diff
@@ -2202,11 +2211,19 @@ void FunctionDefinitionWriter::visit(AST::PhonyAssignmentStatement& statement)
         visit(*statement.maybeExpression());
     }
+
+    if (m_cachedEntryPointStage == ShaderStage::Fragment && m_callGraph.ast().usesFragDepth()) {
```
Same as above, maybe there is a much better way to achieve this, I defer to @tadeuzagallo

mwyrzykowski commented on Source/WebGPU/WebGPU/RenderPassEncoder.mm:
```diff
@@ -138,6 +138,22 @@
+
+    switch (pixelFormat) {
+    case WGPUTextureFormat_Depth16Unorm:
+        return depthClearValue + 1.0 / USHRT_MAX;
```
This might be a Metal driver bug, I need to investigate more closely. But I think more accurately it is because 0.5 can not be representing as a unorm with 16-bit precision

djg replied:
What's the issue? (Yes, 0.5 isn't exactly representable)

mwyrzykowski replied:
if I write 0.5 to clearValue, then it comes across as 0.499992 for Depth16Unorm depth target.

If 0.5 is written out from the fragment shader to the depth value, and depth comparison function of `MTLCompareFunctionNotEqual` is used, then it is determined that 0.5 != 0.499992 (the value in the depth target) so the write occurs when it should not.

Adding 1.0 / uint16_max for the clear value gets a value of '0.5' in the tools debugger for Depth16Unorm and then using `MTLCompareFunctionNotEqual` later does not commit the color write when the depth is written out as 0.5.

It really seems like a driver bug or at least an inconsistency. Setting 0.5 as the depth clear value and writing 0.5 out from the fragment shader to the depth should result in the same value.

mwyrzykowski replied:
I also tried 0.5 / uint16_max thinking something was rounding up when it should not. But the depth clear value was still too low (less than 0.5).

If I can make a repro case I will report a bug

djg commented on Source/WebGPU/WebGPU/CommandEncoder.mm:
```diff
@@ -352,6 +352,69 @@ static auto timestampWriteIndex(auto writeIndex)
+    options.fastMathEnabled = YES;
+ALLOW_DEPRECATED_DECLARATIONS_END
+    id<MTLDevice> device = m_device->device();
+    id<MTLLibrary> library = [device newLibraryWithSource:[NSString stringWithFormat:@"using namespace metal;\n[[vertex]] float4 vsMin(unsigned int vid [[vertex_id]]) { float d = %f; float4 v[] = { float4(-1,-1,d,1), float4(-1,4,d,1), float4(4,-1,d,1) }; return v[vid]; }\n[[vertex]] float4 vsMax(unsigned int vid [[vertex_id]]) { float d = %f; float4 v[] = { float4(-1,-1,d,1), float4(-1,4,d,1), fl ...
```
Could we pull this source string out into a static constant R"()" string literal to make it easier to read?

What's the significance of the 4 in `float4(-1, 4, f, 1)`? And does `vid` need to be modulo 3?

mwyrzykowski replied:
I only draw 1 triangle.

It's a triangle which covers the entire viewport post transform. Do I have my math wrong (very possible)? The post-transform cube has bounds [-1,1] in x and y directions so I draw a triangle like this:
```
|\
| \
|  \
|   \
|    \
|     \
|      \
|       \
|-------|\
|       | \
|       |  \
|-------|___\
```
let me re-run the math on that square right now

mwyrzykowski replied:
oh good catch, the 4 should be a 3

mwyrzykowski replied:
in practice it won't matter since the triangle will be a little too large

mwyrzykowski replied:
I will make it an `"R("` string good idea.

The idea on the z-component is that only values less than min depth and values greater than max depth should pass the depth test. It is used in combination with the depth comparison functions.

mwyrzykowski replied:
@djg so this whole extra pass was completely unnecessary. I added it before I added the WGSL code generation, thinking we could do without extra instructions in the generated WGSL code. However, the CTS revealed it was inadequate and the code-gen was needed.

With the extra codegen, we don't need the extra render pass, so I just removed it.

djg commented on Source/WebGPU/WebGPU/CommandEncoder.mm:
```diff
@@ -584,7 +647,8 @@ static bool isRenderableTextureView(const TextureView& texture)
         if (hasDepthComponent) {
             const auto& mtlAttachment = mtlDescriptor.depthAttachment;
-            mtlAttachment.clearDepth = std::min(1.f, std::max(0.f, attachment->depthClearValue));
+            auto clearDepth = std::min(1., std::max(0., RenderPassEncoder::quantizedDepthValue(attachment->depthClearValue, textureView.format())));
```
`std::clamp` (https://en.cppreference.com/w/cpp/algorithm/clamp)

djg reviewed (commented):
I defer to @tadeuzagallo on the compiler implementation changes.

tadeuzagallo commented on Source/WebGPU/WGSL/Metal/MetalFunctionWriter.cpp:
```diff
@@ -468,6 +472,7 @@ void FunctionDefinitionWriter::visit(AST::Function& functionDefinition)
         first = false;
     }
     // Clear the flag set while serializing StageAttribute
+    m_cachedEntryPointStage = m_entryPointStage;
```
Maybe you could use `m_entryPointStage` and just wait until after you visit the function's body to clear it?

tadeuzagallo commented on Source/WebGPU/WGSL/Metal/MetalFunctionWriter.cpp:
```diff
@@ -2202,11 +2211,19 @@ void FunctionDefinitionWriter::visit(AST::PhonyAssignmentStatement& statement)
     }
+
+    if (m_cachedEntryPointStage == ShaderStage::Fragment && m_callGraph.ast().usesFragDepth()) {
+        m_stringBuilder.append(";\n__wgslFragmentReturnResult.", m_depthAttributeMemberName, " = clamp(__wgslFragmentReturnResult.", m_depthAttributeMemberName, ", as_type<float>(__DynamicOffsets[0]), as_type<float>(__DynamicOffsets[1]));\n");
```
Won't this break if you have more than one fragment shader in the same source? I think we might need to look up the current function's return type, then traverse the struct and find the correct member.

mwyrzykowski replied:
Oh good point, since they might not all write depth. I will try your suggestion, thank you!

mwyrzykowski commented on Source/WebGPU/WGSL/Metal/MetalFunctionWriter.cpp:
```diff
@@ -470,12 +471,16 @@ void FunctionDefinitionWriter::visit(AST::Function& functionDefinition)
         }
         first = false;
     }
+
```
I can remove this added line

mwyrzykowski commented on Source/WebGPU/WGSL/Metal/MetalFunctionWriter.cpp:
```diff
@@ -2227,13 +2232,51 @@ void FunctionDefinitionWriter::visit(AST::PhonyAssignmentStatement& statement)
 
+static std::optional<std::pair<String, String>> fragDepthIdentifierForFunction(AST::Function* function)
+{
+    if (!function || function->stage() != ShaderStage::Fragment)
```
@tadeuzagallo this seems to work and should be local to the current function, what do you think?

tadeuzagallo replied:
Looks good. I was concerned if frag_depth could be nested further in another struct, but it seems like that's not allowed.

mwyrzykowski committed:
[WebGPU] Clamp depth values to the viewport range
https://bugs.webkit.org/show_bug.cgi?id=264449
<radar://118144869>

Reviewed by Tadeu Zagallo.

Add passing expectations for api,operation,rendering,depth_clip_clamp
and add depth clamping when the viewport is not the default [0, 1] range.

* LayoutTests/http/tests/webgpu/webgpu/api/operation/rendering/depth_clip_clamp-expected.txt:
* Source/WebGPU/WGSL/GlobalVariableRewriter.cpp:
(WGSL::RewriteGlobalVariables::insertParameters):
* Source/WebGPU/WGSL/Metal/MetalFunctionWriter.cpp:
(WGSL::Metal::FunctionDefinitionWriter::visit):
* Source/WebGPU/WebGPU/CommandEncoder.h:
* Source/WebGPU/WebGPU/CommandEncoder.mm:
(WebGPU::CommandEncoder::runDepthClampEncoder):
(WebGPU::CommandEncoder::beginRenderPass):
* Source/WebGPU/WebGPU/RenderPassEncoder.h:
* Source/WebGPU/WebGPU/RenderPassEncoder.mm:
(WebGPU::RenderPassEncoder::RenderPassEncoder):
(WebGPU::RenderPassEncoder::quantizedDepthValue):
(WebGPU::RenderPassEncoder::setPipeline):
(WebGPU::RenderPassEncoder::setViewport):

Canonical link: https://commits.webkit.org/274666@main
